app/models/ppm_estimator.rb

```diff
@@ -0,0 +1,212 @@
+  end
+
+  def advance_range
+    @advance_range ||= (total_incentive_range.min * (advance_percentage / 100.0)).to_i..(total_incentive_range.max * (advance_percentage / 100.0)).to_i
+  end
+
+  def valid?
+    # weight_progear and weight_progear_spouse can be empty, will be regarded as 0
+    required_params = %w[rank branch dependents married start end date weight selfpack]
+
+    estimator_params.permitted? &&
+      (required_params - estimator_params.keys).empty? &&
+      /^\d{5}$/.match?(estimator_params[:start]) &&
+      /^\d{5}$/.match?(estimator_params[:end])
+  end
+
+  private
+
+  def estimator_params
+    @estimator_params ||= @params.permit(:rank, :branch, :dependents, :married, :start, :end, :date, :weight, :weight_progear, :weight_progear_spouse, :selfpack)
+  end
+
+  def cwt
+    full_weight / 100
+  end
+
+  def entitlement
+    @entitlement ||= Entitlement.find_by(slug: estimator_params[:rank])
+  end
+
+  def entitlement_self
+    estimator_params[:dependents] == 'yes' && entitlement.total_weight_self_plus_dependents ? entitlement.total_weight_self_plus_dependents : entitlement.total_weight_self
+  end
+
+  def entitlement_progear
+    entitlement.pro_gear_weight || 0
+  end
+
+  def entitlement_progear_spouse
+    estimator_params[:married] == 'yes' && entitlement.pro_gear_weight_spouse ? entitlement.pro_gear_weight_spouse : 0
+  end
+
+  def weight_self_limited
+    [estimator_params[:weight].to_i, entitlement_self].min
+  end
+
+  def weight_progear_limited
+    [estimator_params[:weight_progear].to_i, entitlement_progear].min
+  end
+
+  def weight_progear_spouse_limited
+    [estimator_params[:weight_progear_spouse].to_i, entitlement_progear_spouse].min
+  end
+
+  def full_pack_cost
+    @full_pack_cost ||= FullPack.rate(effective_400ng_date, orig_svc_area.services_schedule, full_weight) * cwt
+  end
+
+  def orig_svc_area
+    @orig_svc_area ||= ServiceArea.from_date_and_svc_area_number(effective_400ng_date, start_zip3.service_area)
+  end
+
+  def dest_svc_area
+    @dest_svc_area ||= ServiceArea.from_date_and_svc_area_number(effective_400ng_date, end_zip3.service_area)
+  end
+
+  def distance
+    @distance ||= DtodZip3Distance.dist_mi(start_zip3.zip3, end_zip3.zip3).to_i
+  end
+
+  def effective_400ng_date
+    # If the user enters a date that's before or after all 400NG rates in the database, then
+    # use the 400NG rates with an effective date closest to the date the user gave
+    # the FullUnpacks table is the shortest, so should be the quickest to query
+    @effective_400ng_date ||=
+      if (effective_range = FullUnpack.effective_date_range).cover?(date)
+        date
+      elsif date < effective_range.min
+        effective_range.min
+      else
+        effective_range.max
+      end
+  end
+
+  def effective_tdl_date
+    # If the user enters a date that's before or after the TDLs in the database, then
```

Comment:
"Tonnage distribution list." I asked why it got this name, and it has something to do with moving freight, which HHG kind of is.

In practice, a TDL is a period of time, where the TSP ranks do not change. The BVS (best value score) is calculated for TSPs based on the discounts they submitted for the year (rate filing) and their performance during a previous performance period (roughly a few months long), as determined by the results of the CSS (customer satisfaction surveys). These scores are used during the next TDL to decide which TSP gets awarded moves by transporters working at JPPSOs.

For PPMs, the incentive is based on what it would have cost for the top rated TSP for that channel to move the same weight. I use the move date to choose the TDL (a date range), and then look up the discount for that channel.